stdlib: Add SE mode support to multi-program workloads

sangjae4309 opened the pull request:
This PR implements adding multi-prgoram support of stdlib SE mode(#1960).

A new function `set_se_multi_binary_workload()` is added, It distributes each binary and its corresponding arguments to cores.

sangjae4309 committed:
stdlib: Add SE mode support to multi-program workloads

sangjae4309 committed:
stdlib: Add SE mode support to multi-program workloads

powerjg reviewed (commented):
See https://github.com/gem5/gem5/issues/1960#issuecomment-2627744564 for some feedback.

sangjae4309 committed:
stdlib: seperate multi-program function

sangjae4309 commented:
@powerjg 
Thank you for your feedback. I’ve taken your suggestion and implemented the separation of the multi-program function.

However, I’m a bit confused. Could you please confirm if you were referring to `Workload` in `src/python/gem5/resources/workload.py`? I have tested it, and it seems to be running correctly. Here is the script I used for verification: [simple-se-workload-resource.zip](https://github.com/user-attachments/files/18628126/simple-se-workload-resource.zip). And also uploaded its raw code at #1960.

I appreciate your feedback. Looking forward to your response!

powerjg commented:
That looks good. Yes, we need to make sure that workloads (e.g., https://resources.gem5.org/resources/[x86-npb-bt-size-s-run](https://resources.gem5.org/resources/x86-npb-bt-size-s-run/raw?database=gem5-resources&version=1.0.0)/raw?database=gem5-resources&version=1.0.0) will work correctly with `set_se_multi_binary_workload`.

This can be done in another PR. @Harshil2107 making sure `set_se_multi_binary_workload` works as a workload is required for gem5 v25.0. I also think it would be a good idea to factor out the common code between that function and `set_se_binary_workload`.

Harshil2107 commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -61,13 +61,116 @@ class SEBinaryWorkload:
+            process = Process(pid=100 + i)
+            binary_path = binary.get_local_path()
+            process.executable = binary_path
+            process.cmd = [binary_path] + arguments[i]
```
This line throws an index out of bounds error when running binaries like:

```python
board.set_se_multi_binary_workload(
    binaries=[
        obtain_resource("x86-hello64-static"),
        obtain_resource("x86-npb-is-size-s"),
    ],
)
```

The issue arises because we expect the `arguments` list to have the same size as the `binaries` list. I suggest adding a check similar to the one ensuring the number of binaries matches the number of cores.

To improve flexibility, I would suggest making `arguments` optional:  
- If `arguments` is an empty list, default to `process.cmd = [binary_path]`.  
- If `arguments` is non-empty, enforce that its size matches the number of binaries.  

This would allow users to omit `arguments` when not needed.

Harshil2107 commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -61,13 +61,116 @@ class SEBinaryWorkload:
+        # Abstract board. This function will not work otherwise.
+        assert isinstance(self, AbstractBoard)
+        # The number of workload should match the nubmer of cores
+        assert len(binaries) == self.get_processor().get_num_cores()
```
It would be helpful for the user to understand why the assertion is failing. Right now, the program only prints `AssertionError: <EMPTY MESSAGE>`. Adding a descriptive error message would make it clearer what went wrong and how to fix it.

sangjae4309 committed:
stdlib: trim and fcator out multi_workload function

sangjae4309 commented:
@powerjg
Ah, I got it!!. I will run another workloads and explore improvement in a future PR.

@Harshil2107 
Thank you for your feedback. I have incorporated your suggestions into my code and also refactored the common logic with `set_se_binary_workload`.

Looking forward to your response!

sangjae4309 committed:
Merge branch 'develop' into stdlib/multi-program

sangjae4309 reviewed (commented):
.

Harshil2107 commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -61,13 +61,145 @@ class SEBinaryWorkload:
+
+        multiprocesses = []
+        for i, binary in enumerate(binaries):
+            process = self._create_process(
```
I think `arguments` should also be passed here as a parameter, as otherwise we would not be able to run the binaries that require arguments.

I would also recommend something along the lines of 

```python
binary_arguments=[]
if arguments:
    binary_arguments=arguments[i]
process = self._create_process(
                binary=binary,
                pid=100 + i,
                stdin_file=stdin_file,
                stdout_file=stdout_file,
                stderr_file=stderr_file,
                env_list=env_list,
                arguments=binary_arguments
            )
```

This way if the user is running binaries that do not require arguments, they can just call

```
board.set_se_multi_binary_workload(binaries=[b1,b2])
```

sangjae4309 replied:
Oh, I overlooked the case where an argument is not required. Thanks for pointing it out!

Harshil2107 commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -61,13 +61,145 @@ class SEBinaryWorkload:
+        assert isinstance(self, AbstractBoard)
+
+        num_cores = self.get_processor().get_num_cores()
+        assert len(binaries) == num_cores == len(arguments), (
+            f"Number of binaries({len(binaries)}), "
+            f"arguments({len(arguments)}) and cores({num_cores}) does not match."
+        )
```
```suggestion
        assert len(binaries) == num_cores, (
            f"Number of binaries({len(binaries)}) "
            f"and cores({num_cores}) should be the same."
        )
        if arguments:
            assert len(binaries) == len(arguments), (
                f"Mismatch between binaries and arguments: "
                f"{len(binaries)} binaries but {len(arguments)} argument lists provided. "
                "The 'arguments' list is optional if none of the binaries require arguments. "
                "However, if any binary requires arguments, a list must be provided for all binaries. "
                "For binaries that do not require arguments, an empty list should be used."
)
        
```

I would suggest updating the assert to something similar to what I mentioned above. As the `arguments` list might be empty if none of the binaries need any arguments. 

This is based on my understanding of the functionality we want. Below is an example of how I think the use case is:

Lets say we have 3 binaries: `b1`, `b2` and `b3`. Only binary `b3` needs an argument `arg1`.

we can call `board.set_se_multi_binary_workload(binaries=[b1,b2])` and `board.set_se_multi_binary_workload(binaries=[b1,b2,b3], arguments=[[],[],[arg1]])` but we cannot call `board.set_se_multi_binary_workload(binaries=[b1,b2,b3], arguements=[[arg1]])`.

Please let me know if I am misunderstanding something.

sangjae4309 replied:
Right, that is exactly what I intended. If there is at least one argument, the sequence count must be properly aligned.

sangjae4309 replied:
Thanks for adding the detailed descriptions. That is  very helpful to understand.

Harshil2107 commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -109,29 +241,17 @@ def set_se_binary_workload(
-            process.errout = stderr_file.as_posix()
-        if env_list is not None:
-            process.env = env_list
+        process = self._create_process(
+            binary=binary,
+            pid=100,
+            stdin_file=stdin_file,
+            stdout_file=stdout_file,
+            stderr_file=stderr_file,
+            env_list=env_list,
+        )
```
```suggestion
        process = self._create_process(
            binary=binary,
            pid=100,
            stdin_file=stdin_file,
            stdout_file=stdout_file,
            stderr_file=stderr_file,
            env_list=env_list,
            arguments=arguments,
        )
```

The parameter `arguments` should also be passed otherwise we wont we able to run binaries that require arguments.

sangjae4309 replied:
Oh, I missed the parameter. Thanks for pointing it out.

Harshil2107 commented:
Thank you for the changes, The `_create_process` function looks good to me. I have requested a few  more small changes, mainly related to the `arguments` parameter.

sangjae4309 committed:
Merge branch 'develop' into stdlib/multi-program

sangjae4309 committed:
Update src/python/gem5/components/boards/se_binary_workload.py

Co-authored-by: Harshil Patel <harshilp2107@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

sangjae4309 committed:
stdlib:Add SE mode support to multi-program workloads

sangjae4309 commented:
@Harshil2107 
Thanks for pointing several bugs and adding helpful suggestion. I have made the requested changes.

sangjae4309 committed:
Merge branch 'develop' into stdlib/multi-program

Harshil2107 reviewed (dismissed):
lgtm

sangjae4309 committed:
Merge branch 'develop' into stdlib/multi-program

powerjg commented on src/python/gem5/components/boards/se_binary_workload.py:
```diff
@@ -61,13 +61,157 @@ class SEBinaryWorkload:
+        self,
+        binaries: List[BinaryResource],
+        exit_on_work_items: bool = True,
+        stdin_file: Optional[FileResource] = None,
+        stdout_file: Optional[Path] = None,
+        stderr_file: Optional[Path] = None,
```
If you're running multiple binaries, I believe you will need to have all of these parameters also be a `List`. Especially for `stdin` you're going to want one per binary. Generally, I think you want a different file descriptor for all `stdin`,  `stdout`, and `stderr` for each binary.

sangjae4309 replied:
Thank you for pointing out!! I've implemented your suggestion.

sangjae4309 committed:
Merge branch 'develop' into stdlib/multi-program

sangjae4309 committed:
stdlib: break arguments into pieces

sangjae4309 committed:
stdlib:factor dummy code out

sangjae4309 commented:
Thanks for the review! I've identified some mistakes in the code and have implemented the necessary corrections. I’d appreciate any further review.

powerjg reviewed (commented):
Thanks for all of the updates! This looks good to me, but I'll let @Harshil2107 sign off.

Harshil2107 reviewed (approved):
lgtm

sangjae4309 committed:
Merge branch 'develop' into stdlib/multi-program
